File: Taboola SDK ObjC Samples/Taboola SDK ObjC Samples/SDK Standerd/UIScrollView/MidArticleManualScrollViewController/MidArticleManualScrollViewController.m
//
//  MidArticleManualScrollViewController.m
//  Taboola SDK ObjC Samples
//
//  Created by Roman Slyepko on 1/28/19.
//  Copyright © 2019 Taboola. All rights reserved.
//

#import "MidArticleManualScrollViewController.h"
#import <TaboolaSDK/TaboolaSDK.h>

@interface MidArticleManualScrollViewController () <UIScrollViewDelegate, TaboolaViewDelegate>
@property (weak, nonatomic) IBOutlet UIScrollView *scrollView;
@property (weak, nonatomic) IBOutlet TaboolaView *midTaboolaView;
@property (weak, nonatomic) IBOutlet UILabel *topLabel;
@property (weak, nonatomic) IBOutlet UILabel *midLabel;
@property (weak, nonatomic) IBOutlet TaboolaView *feedTaboolaView;
@end

@implementation MidArticleManualScrollViewController

- (void)viewDidLoad {
    [super viewDidLoad];
//    self.scrollView.delegate = self;
    //load tabolaView
    self.midTaboolaView.delegate = self;
    self.midTaboolaView.ownerViewController = self;
    self.midTaboolaView.publisher = @"sdk-tester";
    self.midTaboolaView.mode = @"alternating-widget-without-video-1-on-1";
    self.midTaboolaView.pageType = @"article";
    self.midTaboolaView.pageUrl = @"http://www.example.com";
    self.midTaboolaView.placement = @"Mid Article";
    self.midTaboolaView.targetType = @"mix";
    self.midTaboolaView.logLevel = LogLevelDebug;
    [self.midTaboolaView fetchContent];
    
    self.feedTaboolaView.delegate = self;
    self.feedTaboolaView.ownerViewController = self;
    self.feedTaboolaView.publisher = @"sdk-tester";
    self.feedTaboolaView.mode = @"thumbs-feed-01";
    self.feedTaboolaView.pageType = @"article";
    self.feedTaboolaView.pageUrl = @"http://www.example.com";
    self.feedTaboolaView.placement = @"Feed without video";
    self.feedTaboolaView.targetType = @"mix";
    self.feedTaboolaView.logLevel = LogLevelDebug;
    [self.feedTaboolaView setOverrideScrollIntercept:YES];
    [self.feedTaboolaView fetchContent];
}

-(void)dealloc {
    [_midTaboolaView reset];
    [_feedTaboolaView reset];
}

#pragma mark - TaboolaViewDelegate

- (void)taboolaView:(UIView *)taboolaView didLoadPlacementNamed:(NSString *)placementName withHeight:(CGFloat)height {
    NSLog(@"%@", placementName);
}

- (void)taboolaView:(UIView *)taboolaView didFailToLoadPlacementNamed:(NSString *)placementName withErrorMessage:(NSString *)error {
    NSLog(@"%@", error);
}

-(BOOL)onItemClick:(NSString *)placementName withItemId:(NSString *)itemId withClickUrl:(NSString *)clickUrl isOrganic:(BOOL)organic {
    return YES;
}

- (void)scrollViewDidScrollToTopTaboolaView:(UIView*)taboolaView {
    if (self.feedTaboolaView.scrollEnable) {
        NSLog(@"did finish scrolling taboola");
        [self.feedTaboolaView setScrollEnable:NO];
        self.scrollView.scrollEnabled = YES;
    }
    
}

#pragma mark - UIScrollViewDelegate

- (void)scrollViewDidScroll:(UIScrollView *)scrollView {
    NSLog(@"scrollViewDidScroll! %f",scrollView.contentOffset.y);
    [self didEndScrollOfParentScroll];
}

-(BOOL)didEndScrollOfParentScroll {
    float h = self.scrollView.contentSize.height - self.scrollView.bounds.size.height;
    float inset = 0;
    if (@available(iOS 11.0, *)) {
        inset = self.scrollView.adjustedContentInset.bottom;
    } else {
        inset = self.scrollView.contentInset.bottom;
    }
    float y = self.scrollView.contentOffset.y - inset;
    if (h <= y && self.scrollView.scrollEnabled && self.scrollView.contentSize.height > TaboolaView.widgetHeight) {
        NSLog(@"did finish scrolling");
        [self.feedTaboolaView setScrollEnable:YES];
        self.scrollView.scrollEnabled = NO;
        return YES;
    }
    return NO;
}



@end

File: Taboola_SDK_Swift_Sample/Taboola SDK Swift Sample/Taboola SDK Swift Sample/SDK Standard/UIScrollView/MidArticleManualScrollViewController/MidArticleManualScrollViewController.swift
//
//  MidArticleManualScrollViewController.swift
//  Taboola SDK Swift Sample
//
//  Created by Roman Slyepko on 3/13/19.
//  Copyright © 2019 Taboola LTD. All rights reserved.
//

import UIKit
import TaboolaSDK

class MidArticleManualScrollViewController: UIViewController {
    @IBOutlet weak var scrollView: UIScrollView!
    @IBOutlet weak var midTaboolaView: TaboolaView!
    @IBOutlet weak var feedTaboolaView: TaboolaView!
    @IBOutlet weak var topLabel: UILabel!
    @IBOutlet weak var midLabel: UILabel!
    
    var didLoadFeed = false
    
    lazy var viewId: String = {
        let timestamp = Int(Date().timeIntervalSince1970)
        return "\(timestamp)"
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        //load mid tabolaView
        midTaboolaView.delegate = self
        midTaboolaView.ownerViewController = self
        midTaboolaView.publisher = "sdk-tester"
        midTaboolaView.mode = "alternating-widget-without-video-1-on-1"
        midTaboolaView.pageType = "article"
        midTaboolaView.pageUrl = "http://www.example.com"
        midTaboolaView.placement = "Mid Article"
        midTaboolaView.targetType = "mix"
        midTaboolaView.logLevel = .debug
        midTaboolaView.viewID = viewId;
        midTaboolaView.fetchContent()
        
        // load feed tabolaView
        feedTaboolaView.delegate = self
        feedTaboolaView.ownerViewController = self
        feedTaboolaView.publisher = "sdk-tester"
        feedTaboolaView.mode = "thumbs-feed-01"
        feedTaboolaView.pageType = "article"
        feedTaboolaView.pageUrl = "http://www.example.com"
        feedTaboolaView.placement = "Feed without video"
        feedTaboolaView.targetType = "mix"
        feedTaboolaView.logLevel = .debug
        feedTaboolaView.overrideScrollIntercept = true
        feedTaboolaView.viewID = viewId
        
    }
    
    deinit {
        midTaboolaView.reset()
        feedTaboolaView.reset()
    }
    
}

extension MidArticleManualScrollViewController: TaboolaViewDelegate {
    
    func taboolaView(_ taboolaView: UIView!, didLoadPlacementNamed placementName: String!, withHeight height: CGFloat) {
        if placementName == "Mid Article" {
            if !didLoadFeed {
                didLoadFeed = true
                // We are loading the feed only when the widget finished loading- for dedup.
                feedTaboolaView.fetchContent()
            }
        }
        print("Did load: \(placementName)")
    }
    
    func taboolaView(_ taboolaView: UIView!, didFailToLoadPlacementNamed placementName: String!, withErrorMessage error: String!) {
        print("Did fail: \(placementName) error: \(error)")
    }
    
    func onItemClick(_ placementName: String!, withItemId itemId: String!, withClickUrl clickUrl: String!, isOrganic organic: Bool) -> Bool {
        return true
    }
    
    func scrollViewDidScroll(toTopTaboolaView taboolaView: UIView!) {
        if feedTaboolaView.scrollEnable {
            print("did finish scrolling taboola")
            feedTaboolaView.scrollEnable = false
            scrollView.isScrollEnabled = true
        }
    }
}


extension MidArticleManualScrollViewController: UIScrollViewDelegate {
    func scrollViewDidScroll(_ scrollView: UIScrollView) {
        print("scrollViewDidScroll: \(scrollView.contentOffset.y)")
        _ = didEndScrollOfParrent()
    }
    
    func didEndScrollOfParrent() -> Bool {
        let height = scrollView.contentSize.height - scrollView.bounds.size.height
        var inset: CGFloat = 0.0
        if #available(iOS 11.0, *) {
            inset = scrollView.adjustedContentInset.bottom
        } else {
            inset = scrollView.contentInset.bottom
        }
        let y = scrollView.contentOffset.y - inset
        
        if height <= y && scrollView.isScrollEnabled && scrollView.contentSize.height > TaboolaView.widgetHeight() {
            print("did finish scrolling")
            feedTaboolaView.scrollEnable = true
            scrollView.isScrollEnabled = false
            return true
        }
        return false
    }
}
